### ISSUE TITLE
Inclusion to F-Droid and fmod.jar

### ISSUE THREAD

**Issue by ClockGen** [mkrupczak3/Freedoom-for-Android#40]
Hello, is it possible to remove fmod.jar? As the original gzdoom port by nvllsvm, inclusion of this project to F-Droid is blocked by usage of a proprietary library fmod.jar.

**Comment by mkrupczak3** [comment]
Hi ClockGen,

I'm _pretty_ sure that fmod.jar could be safely removed from this app without any issues.

I think that fmod [was used as the default midi device options](https://zdoom.org/wiki/FMOD) in older versions of GZDoom such as [the one used in this app](https://github.com/mkrupczak3/GZDoom-df1364e-link). When porting [nvllsvm's version](https://github.com/nvllsvm/GZDoom-Android) though, I added in a hack to [change the midi device to timidity](https://github.com/mkrupczak3/Freedoom-for-Android/commit/f3b97b50a502880a3048ec5c2ae846f58e1b9e33#r36257099) anyways, so fmod.jar could probably be safely deleted unless it breaks the build

I'm not currently at a place where I have an environment to build and test an updated version of the app, but if anyone is willing to test it building and working without fmod.jar then I'd be willing to accept a pull request to remove it.

Otherwise, I think it would be neat to get a more compliant version of this app on F-Droid. I think it would be great to get more eyes on it and more contributors involved for things like [cool and new features](https://matthew.krupczak.org/2019/10/20/hawking-my-projects-ii-500000-installs-with-freedoom-for-android/). 

Context:
https://forum.f-droid.org/t/freedoom/1913

**Comment by mkrupczak3** [comment]
Also @ClockGen if I don't reply back to an email you send, just assume that I've missed it and feel free to submit a GitHub issue or comment.

**Comment by EmberHeartshine** [comment]
As a stopgap, I can include the releases in my [personal F-Droid repo](https://fdroid.heartshine.xyz/) if that's okay with you; I have a few hundred folks pinging it a day.

**Comment by mkrupczak3** [comment]
@EmberHeartshine If you're willing, that would be cool. 

In addition, people might find it useful to download the releases I have here on Github since they are more up-to-date than those on the Play Store due to #41 

I think the issues brought up by F-droid are pretty important though. Chiefly, that fmod.jar and the like are pre-built, non-free binaries. I think it's a bit of a faux-pas to package and include this app in F-droid before these binaries can be removed, as some users of F-droid could have heightened security and code scrutiny concerns.  

If there was a better way of informing users up front what they're getting "under the hood" I think that would be okay. 

Eventually, I think most of these non-free binaries could be removed without much impact on the app though. I would like to have this happen for the health of the project. 

Unfortunately, the current state of this project is pretty misshapen, and I'm unlikely to be able to make much progress in the near future without more external contributions or a rebase/donated code from my upstream's upstream: [beloko / delta touch](https://github.com/emileb) @emileb

**Comment by EmberHeartshine** [comment]
@mkrupczak3 
> If there was a better way of informing users up front what they're getting "under the hood" I think that would be okay.

The nice thing about F-Droid is that is has "anti-features" flags; in this case, the UpstreamNonFree flag. In the app, that looks like this:
![Screenshot_2020-05-04-11-49-26](https://user-images.githubusercontent.com/50666308/80986324-6ab9ea00-8dfe-11ea-93de-a52e3467b6b3.png)
(I've added a description and reference to this issue but the metadata hasn't updated in my cache yet.)

**Comment by mkrupczak3** [comment]
@EmberHeartshire 

Oh nice, I like that a lot. 

I think it would be great if you're willing to put the release in your personal F-droid repo, as long as you maintain that notice. 

The updated releases on GitHub have more language translations and newer versions of the Freedoom wads than my Google Play release as well.

However you package it, I also recommend add translations for multiple languages to the F-droid store release if such functionality is available

Please let me know if you need a signed app release or anything.

**Comment by EmberHeartshine** [comment]
@mkrupczak3 The update script I've written pulls the release APKs from GitHub, so the way they're being updated now is fine. I'm not actually building them from source myself.

I'll have to look into localization; the way I'm handling metadata is a little outdated so it doesn't support multiple languages but I can fix that.

**Comment by ClockGen** [comment]
@mkrupczak3 any progress on this?

**Comment by mkrupczak3** [comment]
@ClockGen No, sorry. I don't think I'll be able to make any progress on this going forward, either. 

My Windows laptop where I had this entire Android Studio project on got wiped when I installed Linux a while back. Also, the build tools necessary to develop this app were completely deprecated by @Google for no good reason.

Developing this app has been painful, but a very good experience. I know now to never rely on the Android ecosystem for a livelyhood, because there is absolutely no consistency, stability, or sanity to be found there. 

If you're still interested in building a completely FOSS version of this app, might have to try it yourself. I've left some instructions in some of these GitHub issues, but you would have to set up a janky dev environment to get it done. 

Also worth checking out, hound @emileb to see if he'd be willing to make a limited version of his "Delta Touch" android app as open source code. 

It's unclear if his use of the GPL'd GZDoom engine *technically* obligates him to release a limited version of his launcher code to customers per the terms of the GPL, but it seems a bit unfair to make oodles of money off of open source code without giving a bit back.

**Comment by emileb** [comment]
Hi guys, sorry I didn't replay to this before, thanks for the ping! I am updating your last Freedoom-for-Android:

https://github.com/emileb/Freedoom-for-Android/commits/master

It now builds with latest Android Studio and NDK 21.0 ( I haven't updated to r22 yet) and I did a couple of fixes to the startup of the app.

I have removed the patching stuff and it now uses my copy of GZDoom 1.9 - the advantage of this versions is it includes OpenAL  as an option for sound backend so FMOD can be removed.

Happy to make a branch with FMOD gone, is that the only stopping addition to F-Droid now?

**Comment by ClockGen** [comment]
@emileb most likely yes. I've been fixing it up on my own and got reproducible builds and started patching proprietaty stuff out. I patched out com.bda.controller library from your touchcontrols lib, since it's proprietary. I take it's used for hardware controller support, so, unless there's some other way, no hardware controller support for f-droid. But FMOD is more important, because so far my attempts to remove it produced builds that crash on startup and I'm in no environment to debug it.

**Comment by emileb** [comment]
Ah yes that Moga controller stuff needs to go also. Well feel free to give my copy a go, it builds for me on Windows - you will need to remove the patching commands from build.sh if using that. 

Once you have that building I can remove the FMOD stuff. The issue with the build you have is if you remove FMOD it will then have no sound!

**Comment by ClockGen** [comment]
@emileb Haha I figured out that much already. I have patches to to remove moga lib already, it builds and works fine, at least with touchscreen, I can submit them as a PR in your repo.

**Comment by emileb** [comment]
Oh yeah BTW hw controller support will still work for 99% without that library, it's just needed for a very old specific Moga controller, Android has built in controller support for most gamepads.

**Comment by mkrupczak3** [comment]
> Hi guys, sorry I didn't replay to this before, thanks for the ping! I am updating your last Freedoom-for-Android:
> 
> https://github.com/emileb/Freedoom-for-Android/commits/master
> 
> It now builds with latest Android Studio and NDK 21.0 ( I haven't updated to r22 yet) and I did a couple of fixes to the startup of the app.
> 
> I have removed the patching stuff and it now uses my copy of GZDoom 1.9 - the advantage of this versions is it includes OpenAL as an option for sound backend so FMOD can be removed.
> 
> Happy to make a branch with FMOD gone, is that the only stopping addition to F-Droid now?

Hey, thanks for the build @emileb. I apologize for the tone of my previous comment, I'm in great debt to your great codebase for making this project possible. 

Excited a the prospect of a truly FOSS GZDoom running on Android. Keep up the great work

**Comment by mkrupczak3** [comment]
Per FMOD, I don't know if its relevant, but I added a .ini file hack to the Freedoom app to change the sound device from the default FMOD ouptut to Timidity. 

If FMOD is essential to the build, it may be possible to replace it with a dummy replacement that doesn't actually do anything. This would not have any impact on the sound output, but would allow the build to succeed.

**Comment by ClockGen** [comment]
It still includes various headers and cpp files from fmod (which are not gpl-licensed) as well as a shared library in build.

**Comment by emileb** [comment]
Hey @mkrupczak3 no it's all good, I should have replied earlier to this! 

OK just in case there was any confusing here - There is FMOD which is the backend sound output device, the setups and open the audio in the version of GZDoom Freedoom currently uses - all audio output goes through this.

Then there are various *music* renderers - Fluidsynth, Timidity, GME, OPL etc, these generate the music data which ultimately goes through FMOD to play.

AFAIK all the music renderers are good to go, they all include the source code. We need to remove FMOD however, but in the current version of GZDoom you have it is the ONLY audio backend so this would break all audio in the app. 

In the update I just did I am using a newer version of GZDoom which also includes FMOD **and OpenAL** as audio backend so we have the option to remove FMOD and use OpenAL instead. It should then behave essentially the same as before (possibly missing some useless audio features like 'Reverb' which are present in FMOD)

I'll look into remove FMOD from my build this week.

Hope that makes sense and sorry if you knew that already, cheers 👍

**Comment by emileb** [comment]
> It still includes various headers and cpp files from fmod (which are not gpl-licensed) as well as a shared library in build.

I've not removed it yet, I wanted to make sure it was in a working state first.

**Comment by ClockGen** [comment]
No, I answered @mkrupczak3, that in the current state of this repo removing only fmod.jar is not enough, there's still native part of fmod present. 

Anyway I'm having problems compiling your version, it seems that some sources are missing. I had to manually initialize some submodule (jwzgles is not listed as submodule), and fix some paths to compile it on linux (different letter cases), but now I'm getting this error:
```
[armeabi-v7a] Compile++ thumb: gzdoom <= parse_xlat.cpp
jni/gzdoom_android/mobile/..//mobile/../src/xlat/parse_xlat.cpp:43:10: fatal error: 'xlat_parser.h' file not found
#include "xlat_parser.h"
```
and this header is nowhere to be found in the whole project tree.

**Comment by emileb** [comment]
Ah yeah sorry! They were included in .gitignore so I missed them.
Try again: https://github.com/emileb/gzdoom/commits/freedoom_for_android_gz1.9

**Comment by ClockGen** [comment]
@emileb is no_fmod branch good to go? I merged it locally and it builds fine, audio seems to work (music and effects). If it is, can you merge it please? I'll start working on F-Droid inclusion then.

**Comment by emileb** [comment]
Yeah I think's good to go as a first pass. 
Actually I was going to ask @mkrupczak3  if he wants to keep a branch going with FMOD/MOGA lib included in case he wanted to use it for Google Play? Or just remove FMOD completely for all builds from now on?

**Comment by mkrupczak3** [comment]
@emileb I'd say let loose and do whatever you can to remove non-free binary blobs.

I can't update the app on Google Play anymore because I haven't figured out how to compile for Armv8

I still feel that a fully FOSS release would provide the most new value to users

**Comment by emileb** [comment]
OK I've merged no_fmod back into master. BTW @mkrupczak3 I should be able to fix it up to build on arm64-v8a, I'll take a look.

**Comment by ClockGen** [comment]
@emileb @mkrupczak3 I've opened a packaging MR on F-Droid gitlab, https://gitlab.com/fdroid/fdroiddata/-/merge_requests/8358, there's a few questions. If you don't have gitlab accounts you can just post here, I'll relay it.

**Comment by emileb** [comment]
com.android.vending.CHECK_LICENSE is a leftover as you say, it can be removed.
BTW does this fdroid build point to my repo or did you do a fork? Just so I know if to be careful and not break it.

**Comment by ClockGen** [comment]
@emileb no, I'm using your repo, all the submodules are yours anyway.

**Comment by ClockGen** [comment]
@emileb I opened another PR with some cleanup, to not cause confusion if some F-Droid users decide to visit the repo.

**Comment by emileb** [comment]
Thanks for the ping, for some reason I'm not getting notifications of pull requests. Now Merged.

**Comment by ClockGen** [comment]
It's merged now, and, hopefully, will appear in the next f-droid build cycle.

**Comment by mkrupczak3** [comment]
Thanks @emileb @ClockGen for making this app truly free and open source. This issue can be closed
